### ISSUE TITLE
Option to cleanly exit a CalcJob before hitting wallclock limit?

### ISSUE THREAD

**Issue by greschd** [aiidateam/aiida-core#3475]
Some codes provide an option to cleanly shut down a calculation, e.g. by writing an ``EXIT`` (or similarly named) file somewhere in the working directory.

I was wondering if AiiDA has a feature / hook for doing this when the run-time of the calcjob approaches the set wallclock limit.

If there is no such a feature currently, I don't see it as a very high priority. It could be tricky to reliably have the daemon execute this at roughly the right time.

**Comment by sphuber** [comment]
No, this feature does not exist and as you say, I think this would be relatively tricky. At launch time, we could schedule an event in the future at which point the wall time will almost expire and have the engine write this file in the working directory, but as you say giving guarantees that this will be done in time under heavy load will be almost impossible. Also of course there will be delays with the transport connection delay which is configurable.

Besides, if a code provides such a hook, doesn't it make a lot more sense to expose this through an input parameter. For example with Quantum ESPRESSO you can specify `max_wallclock_seconds` in the inputs and we use this to prevent abrupt shutdowns by the scheduler.

**Comment by greschd** [comment]
> Besides, if a code provides such a hook, doesn't it make a lot more sense to expose this through an input parameter. For example with Quantum ESPRESSO you can specify `max_wallclock_seconds` in the inputs and we use this to prevent abrupt shutdowns by the scheduler.

That _would_ be the better place to put it (especially since the code might know how long the next "uninterruptible" iteration might take), but changes on the level of the code can be... difficult 😅

It might still be useful to have a hook that _might_ execute before the code hits the wall.
